### PR TITLE
netdata virtual memory size

### PR DESCRIPTION
**Pull request by ktsaou**
This PR adds support for linking with jemalloc and tcmalloc. Both are optional and controlled via `--with-jemalloc` and `--with-tcmalloc` `./configure` options.

## netdata virtual memory size

You may notice that netdata's virtual memory size, as reported by `ps` or `/proc/pid/status`  (or even netdata's applications virtual memory chart)  is unrealistically high.

For example, it may be reported to be 150+MB, even if the resident memory size is just 25MB. Similar values may be reported for netdata plugins too.

Check this for example: A netdata installation with default settings on Ubuntu 16.04LTS. The top chart is **real memory used**, while the bottom one is **virtual memory**:

image

### why this happens?

The system memory allocator allocates virtual memory arenas, per thread running. To my observations, this defaults to 16MB per thread on 64 bit machines. So, if you get the difference between real and virtual memory and divide it by 16MB you will roughly get the number of threads running.

The system does this for speed. Having a separate memory arena for each thread, allows the threads to run in parallel in multi-core systems, without any locks between them.

This behaviour is system specific. For example, the chart above when running netdata on alpine linux (that uses **musl** instead of **glibc**) is this:

image

### can we do anything to lower it?

Since netdata already uses minimal memory allocations while it runs (i.e. it adapts its memory on start, so that while repeatedly collects data it does not do memory allocations), it already instructs the system memory allocator to minimize the memory arenas for each thread. We have also added 2 configuration options to allow you tweak these settings: https://github.com/firehol/netdata/blob/5645b1ee35248d94e6931b64a8688f7f0d865ec6/src/main.c#L410-L418

However, even if we instructed the memory allocator to use just one arena, it seem it allocates an arena per thread.

This PR adds support for `jemalloc` and `tcmalloc`, however both of which behave exactly the same to the default memory allocator in this aspect.

### Is this a problem?

No, it is not.

Linux reserves real memory (physical RAM) in pages (on x86 machines pages are 4KB each). So even if the system memory allocator is allocating huge amounts of virtual memory, only the 4KB pages that are actually used are reserving physical RAM.

### CODE REVIEW

**Comment by ktsaou** [comment]
Added wiki page: https://github.com/firehol/netdata/wiki/netdata-virtual-memory-size

**Comment by ktsaou** [comment]
@alonbl I added jemalloc and tcmalloc support (not really needed, since netdata does not make frequent allocations). Good to have them as alternatives for the future though.

I also used atomic operations when available for the malloc debugging info.

If you know any other memory allocators that can lower the virtual memory size of netdata, let me know...

File: src/common.h [hunk 1]
```diff
@@ -5,6 +5,18 @@
 #include <config.h>
 #endif
 
+#ifdef has_jemalloc
+#include <jemalloc/jemalloc.h>
+#undef HAVE_C_MALLOPT
+#undef HAVE_C_MALLINFO
```
Comment (alonbl) [inline src/common.h]: please do not do this, you can never know what implications are.
if you want define something to skip definition, but most probably you do not want to define them at first in configure.

File: src/common.h [hunk 2]
```diff
@@ -5,6 +5,18 @@
 #include <config.h>
 #endif
 
+#ifdef has_jemalloc
```
Comment (alonbl) [inline src/common.h]: you probably want to define our own ENABLE_JEMALLOC in configure.ac to make sure conventions are set.

File: src/main.c [hunk 3]
```diff
@@ -374,7 +374,7 @@ int main(int argc, char **argv)
                     break;
                 case 'v':
                     // TODO: Outsource version to makefile which can compute version from git.
-                    printf("netdata 1.3.1_master\n");
+                    printf("netdata " VERSION "\n");
```
Comment (alonbl) [inline src/main.c]: you probably want %s and VERSION to avoid assumptions.

File: configure.ac [hunk 4]
```diff
@@ -84,13 +84,19 @@ LIBS="${PTHREAD_LIBS} ${LIBS}"
 CFLAGS="${CFLAGS} ${PTHREAD_CFLAGS}"
 CC="${PTHREAD_CC}"
 
+TS_CHECK_JEMALLOC
+if test "$has_jemalloc" = "0"; then
+	TS_CHECK_TCMALLOC
+fi
+if test "$has_jemalloc$has_tcmalloc" = "00"; then
+	AC_C_MALLOPT
+	AC_C_MALLINFO
```
Comment (alonbl) [inline configure.ac]: you can always check for these, only set ENABLE_something if you want to actually use these feature.

always separate between two processes:
1. detection
2. actual enable for use

**Pull request merged**